File: backend/src/main/java/io/smartbreakdown/services/BreakdownService.java
package io.smartbreakdown.services;

import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.SortedSet;
import java.util.TreeSet;
import java.util.stream.Collectors;

import comparators.BreakdownItemComparator;
import io.smartbreakdown.models.BreakdownItem;
import jakarta.enterprise.context.ApplicationScoped;

@ApplicationScoped
public class BreakdownService {

    private static final List<Double> DENOMINATIONS = Arrays.asList(200.0, 100.0, 50.0, 20.0, 10.0, 5.0, 2.0, 1.0, 0.5,
            0.2, 0.1,
            0.05, 0.02, 0.01);

    private static final BreakdownItemComparator COMPARATOR = new BreakdownItemComparator();
    private static final int CENTS_MULTIPLIER = 100;

    public List<Double> getDenominations() {
        return DENOMINATIONS;
    }

    public BreakdownService() {
    }

    public SortedSet<BreakdownItem> calculateBreakdown(final double amount) {
        validateAmount(amount);

        SortedSet<BreakdownItem> breakdown = new TreeSet<>(COMPARATOR);
        long remainingAmountInCents = convertToCents(amount);

        for (Double denomination : DENOMINATIONS) {
            long denominationInCents = convertToCents(denomination);
            long count = remainingAmountInCents / denominationInCents;

            if (count > 0) {
                breakdown.add(new BreakdownItem(denomination, count));
                remainingAmountInCents = remainingAmountInCents - (count * denominationInCents);
            }
        }

        return breakdown;
    }

    public SortedSet<BreakdownItem> calculateBreakdownDifferences(final SortedSet<BreakdownItem> currentBreakdown,
            final SortedSet<BreakdownItem> previousBreakdown) {
        Map<Double, Long> currentMap = createDenominationMap(currentBreakdown);
        Map<Double, Long> previousMap = createDenominationMap(previousBreakdown);
        Set<Double> affectedDenominations = new HashSet<>();
        affectedDenominations.addAll(currentMap.keySet());
        affectedDenominations.addAll(previousMap.keySet());

        return createDifferencesSet(currentMap, previousMap, affectedDenominations);
    }

    private Map<Double, Long> createDenominationMap(final SortedSet<BreakdownItem> breakdown) {
        return breakdown.stream()
                .collect(Collectors.toMap(
                        BreakdownItem::getDenomination,
                        BreakdownItem::getCount));
    }

    private void validateAmount(final double amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount cannot be negative");
        }
    }

    private long convertToCents(final double amount) {
        return Math.round(amount * CENTS_MULTIPLIER);
    }

    private SortedSet<BreakdownItem> createDifferencesSet(
            Map<Double, Long> currentMap,
            Map<Double, Long> previousMap,
            Set<Double> allDenominations) {

        SortedSet<BreakdownItem> differences = new TreeSet<>(COMPARATOR);

        for (Double denomination : allDenominations) {
            long currentCount = currentMap.getOrDefault(denomination, 0L);
            long previousCount = previousMap.getOrDefault(denomination, 0L);
            long difference = currentCount - previousCount;

            differences.add(new BreakdownItem(denomination, difference));
        }

        return differences;
    }
}

File: frontend/src/services/breakdown/breakdown.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable, of, throwError } from 'rxjs';
import { EURO_DENOMINATIONS } from 'src/constants/denominations';
import { BreakdownItem } from 'src/models/breakdown.model';
import { BreakdownHttpService } from './breakdown-http.service';

export enum CalculationMode {
    FRONTEND,
    BACKEND
}

@Injectable({
    providedIn: 'root'
})
export class BreakdownService {

    constructor(private breakdownHttpService: BreakdownHttpService) { }

    calculateBreakdown(amount: number, calculationMode: CalculationMode = CalculationMode.FRONTEND): Observable<BreakdownItem[]> {
        validateAmount(amount);

        if (calculationMode === CalculationMode.BACKEND) {
            return this.calculateBreakdownBackend(amount);
        } else {
            return this.calculateBreakdownFrontend(amount);
        }
    }

    calculateBreakdownBackend(amount: number): Observable<BreakdownItem[]> {
        return this.breakdownHttpService.calculateBreakdown(amount);
    }

    calculateBreakdownFrontend(amount: number): Observable<BreakdownItem[]> {
        // Working with cents to avoid floating point precision issues
        const breakdown: BreakdownItem[] = [];
        let remainingAmountInCents = Math.round(amount * 100);

        for (const denomination of EURO_DENOMINATIONS) {
            const denominationInCents = Math.round(denomination * 100);
            const count = Math.floor(remainingAmountInCents / denominationInCents);
            if (count > 0) {
                breakdown.push({
                    denomination,
                    count
                });
                remainingAmountInCents = remainingAmountInCents - (count * denominationInCents);
            }
        }
        return of(breakdown);
    }

    calculateBreakdownDifferences(currentBreakdown: BreakdownItem[], previousBreakdown: BreakdownItem[], calculationMode: CalculationMode = CalculationMode.FRONTEND): Observable<BreakdownItem[]> {
        if (calculationMode === CalculationMode.BACKEND) {
            return this.calculateBreakdownDifferencesBackend(currentBreakdown, previousBreakdown);
        } else {
            return this.calculateBreakdownDifferencesFrontend(currentBreakdown, previousBreakdown);
        }
    }
    calculateBreakdownDifferencesBackend(currentBreakdown: BreakdownItem[], previousBreakdown: BreakdownItem[]): Observable<BreakdownItem[]> {
        return this.breakdownHttpService.calculateBreakdownDifferences(currentBreakdown, previousBreakdown);
    }
    calculateBreakdownDifferencesFrontend(currentBreakdown: BreakdownItem[], previousBreakdown: BreakdownItem[]): Observable<BreakdownItem[]> {
        const differences: BreakdownItem[] = [];
        const currentBreakdownMap: Map<number, number> = this.createDenominationMap(currentBreakdown);
        const previousBreakdownMap: Map<number, number> = this.createDenominationMap(previousBreakdown);
        const combinedAffectedDenominations: Set<number> = new Set([...currentBreakdown.map(item => item.denomination), ...previousBreakdown.map(item => item.denomination)]);
        for (const currentDenomination of combinedAffectedDenominations) {
            const previousCount = previousBreakdownMap.get(currentDenomination) || 0;
            const currentCount = currentBreakdownMap.get(currentDenomination) || 0;

            const difference = currentCount - previousCount;
            differences.push({ denomination: currentDenomination, count: difference });
        }
        const sortedDifferences = this.sortByDenomination(differences);
        return of(sortedDifferences);
    }


    private sortByDenomination(items: BreakdownItem[]): BreakdownItem[] {
        return items.sort((a, b) => b.denomination - a.denomination);
    }

    private createDenominationMap(breakdown: BreakdownItem[]): Map<number, number> {
        return new Map<number, number>(breakdown.map(({ denomination, count }) => [denomination, count]));
    }
}

function validateAmount(amount: number) {
    if (amount < 0) {
        throw new Error('Amount cannot be negative');
    }
}
